### ISSUE TITLE
Prefix context in CoarseTransformer and FineTransformer

### ISSUE THREAD

**Issue by zhvng** [lucidrains/audiolm-pytorch#67]
Hey! What is the reason for not using prefix_context for the semantic tokens In CoarseTransformer and the coarse tokens in FineTransformer? It seems that we want to condition on these tokens without needing to calculate the logits, so it could save some computation.

I also noticed the implementation adds the semantic cross-entropy loss to the CoarseTransformer loss and the coarse cross-entropy loss to the FineTransformer loss. I looked through the paper and couldn't seem to find discussion on this. What effect does this have on training?

**Comment by lucidrains** [comment]
I don't think it hurts to have autoregressive loss for the semantic and coarse tokens as this was the same setup for dalle1 for text to image synthesis. I can offer an option to turn off those losses though (and save a tiny bit of compute for their logits)

**Comment by lucidrains** [comment]
if I add the option, maybe you can run the comparison and share your experimental results? Have you been training?

**Comment by zhvng** [comment]
makes sense. I haven't been training, but I've been using the code for another project I'm working on (that is similar to your musiclm-pytorch work). Would be interested in running that comparison afterwards, but it might be a while so no rush in adding that option.

**Comment by lucidrains** [comment]
ok, I'll add it tomorrow (since I'm moving onto protein diffusion soon) along with a bunch of musiclm stuff

**Comment by zhvng** [comment]
sg, thanks! appreciate your work in this space

**Comment by lucidrains** [comment]
@zhvng hey Allen

so i reviewed the code this morning and it seems like i preempted this and have a way to turn off the loss through setting the cross entropy loss weight to 0 for both coarse and fine transformers

however, i also added the option to turn off [the projection of the logits](https://github.com/lucidrains/audiolm-pytorch/commit/8db0afee419423c0040154bd573d19e39ddcb8b0#diff-96a5ee045c1df07f3125d9b4189130620f229785b36cebb86c95b0646f0d744dR562)

let me know if this works for you!

**Comment by zhvng** [comment]
hey Phil, thanks for looking into this! I was more so talking about saving compute in the self attention layer by passing the tokens into the `context` of the transformer and [conditioning as self attention prefix](https://github.com/lucidrains/audiolm-pytorch/blob/main/audiolm_pytorch/audiolm_pytorch.py#L371). Do you think that conditioning the transformer in this way would have a negative impact on performance? I'm now realizing that the self attention on the conditioning tokens might be helping out later layers.

**Comment by lucidrains** [comment]
@zhvng ohh yea, that would for sure impact performance, unless if one were to put the semantic and coarse tokens through its own encoder first (but then it would be largely the same)

**Comment by lucidrains** [comment]
@zhvng actually, i retract my statements, because i recall [Perceiver AR](https://github.com/lucidrains/perceiver-ar-pytorch) using the scheme you described

i'm not sure!

**Comment by zhvng** [comment]
oh hmm that is interesting. I agree with you that that's the safest bet. I might benchmark it sometime and let you know the results.

I also noticed that the model does causal attention over the entire token sequence. Could be interesting to make the attention over the semantic and coarse tokens non-causal wrt themselves

**Comment by lucidrains** [comment]
@zhvng yes! all interesting things to try, would make for a nice paper 😄 

i'm done for today, but will titrate in these different features for you (or someone else) to try

**Comment by zhvng** [comment]
sg! I could help out as well if needed
